File: 백준/Gold/2357. 최솟값과 최댓값/최솟값과 최댓값.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

vector<int> arr, minTree, maxTree;

int minInit(int start, int end, int node) {
    if (start == end) {
        return minTree[node] = arr[start];
    }

    int mid = (start + end) / 2;
    return minTree[node] = min(minInit(start, mid, node * 2), minInit(mid + 1, end, node * 2 + 1));
}

int maxInit(int start, int end, int node) {
    if (start == end) {
        return maxTree[node] = arr[start];
    }

    int mid = (start + end) / 2;
    return maxTree[node] = max(maxInit(start, mid, node * 2), maxInit(mid + 1, end, node * 2 + 1));
}

int minFind(int start, int end, int node, int left, int right) {
    if (right < start || end < left) {
        return INT_MAX;
    }

    if (left <= start && end <= right) {
        return minTree[node];
    }

    int mid = (start + end) / 2;

    return min(minFind(start, mid, node * 2, left, right), minFind(mid + 1, end, node * 2 + 1, left, right));
}

int maxFind(int start, int end, int node, int left, int right) {
    if (right < start || end < left) {
        return INT_MIN;
    }

    if (left <= start && end <= right) {
        return maxTree[node];
    }

    int mid = (start + end) / 2;

    return max(maxFind(start, mid, node * 2, left, right), maxFind(mid + 1, end, node * 2 + 1, left, right));
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, M;
    cin >> N >> M;

    arr.resize(N + 1);
    for (int i = 1; i <= N; i++) {
        cin >> arr[i];
    }

    minTree.resize(N * 4);
    maxTree.resize(N * 4);

    minInit(1, N, 1);
    maxInit(1, N, 1);

    for (int i = 0; i < M; i++) {
        int left, right;
        cin >> left >> right;

        cout << minFind(1, N, 1, left, right) << " " << maxFind(1, N, 1, left, right) << "\n";
    }

    return 0;
}

File: 백준/Gold/2357. 최솟값과 최댓값/최솟값과 최댓값.py
import math
import sys


input = lambda: sys.stdin.readline().rstrip()


def build(node, start, end):
    if start == end:
        tree[node] = (data[start], data[start])
    else:
        mid = (start + end) // 2

        # 왼쪽 생성
        build(node * 2 + 1, start, mid)

        # 오른쪽 생성
        build(node * 2 + 2, mid + 1, end)

        # 값 지정
        tree[node] = (
            min(tree[node * 2 + 1][0], tree[node * 2 + 2][0]), max(tree[node * 2 + 1][1], tree[node * 2 + 2][1]))


def query(left, right, node, start, end):
    if left > end or right < start:
        return (math.inf, -math.inf)

    if left <= start and right >= end:
        return tree[node]

    mid = (start + end) // 2
    left_val = query(left, right, node * 2 + 1, start, mid)
    right_val = query(left, right, node * 2 + 2, mid + 1, end)

    return min(left_val[0], right_val[0]), max(left_val[1], right_val[1])


if __name__ == '__main__':
    N, M = map(int, input().split())
    data = [int(input()) for _ in range(N)]
    tree = [(math.inf, -math.inf) for _ in range(4 * N)]
    build(0, 0, len(data) - 1)

    for _ in range(M):
        a, b = map(int, input().split())
        print(*query(a - 1, b - 1, 0, 0, len(data) - 1))